### PR TITLE
Refactor the Kernel

### PR DESCRIPTION
**Pull request by zebmason**
Refactor the code in Kernels.h & .cpp to C++ code

### CODE REVIEW

**Comment by zebmason** [comment]
This is just a heads up so I don't find anyone else doing the same refactoring

**Comment by CloneDeath** [comment]
Ha, I was actually in the middle of a VERY similar change, before backing out to work on something else.

You have my approval, I look forward to seeing it when done :)

**Comment by zebmason** [comment]
@matt-gretton-dann Are the naming conventions for structs the same as for classes?

**Comment by CloneDeath** [comment]
@matt-gretton-dann Yes, PascalCase for structs as well as classes. And make sure to use snake_case for class/struct functions.

**Comment by zebmason** [comment]
And snake_case_ for public member variables?

**Comment by CloneDeath** [comment]
snake_case_ (with underscore) for private,

I **think** @matt-gretton-dann wanted just snake_case (no underscore) for public, but I can't seem to find where I saw that...

Honestly, I wouldn't stress *too* much about it. If Matt has an issue with something you've done, just make sure to note it in #305 so we can make sure to include it in our style guide.

We're all trying to figure out how things ought to be as a community, you're definitely not alone in this :)

**Comment by zebmason** [comment]
@dlaydon Could you look at the comments in Kernels.h or suggest someone else, please?

File: src/Kernels.cpp [hunk 1]
```diff
@@ -3,132 +3,136 @@
...
-		Kernel = GaussianKernel;
-	else if (P.KernelType == 4)
-		Kernel = StepKernel;
-	else if (P.KernelType == 5)
-		Kernel = PowerKernelB;
-	else if (P.KernelType == 6)
-		Kernel = PowerKernelUS;
-	else if (P.KernelType == 7)
-		Kernel = PowerExpKernel;
+	try
```
Comment (CloneDeath) [inline src/Kernels.cpp:11]: I think this is a case of tell don't ask.

Let the function throw an exception instead of returning a boolean.

Comment (zebmason) [reply src/Kernels.cpp:11]: Did think about that at the time but I'm always loath to change the behaviour of unfamiliar code so I thought I'd do the ugly solution. I'll change it and not do any check on return. Thanks.

**Comment by zebmason** [comment]
@weshinsley Ready for review

File: src/Kernels.h [hunk 2]
```diff
@@ -1,16 +1,122 @@
...
+      /// \return Probability
+      double step(double r2) const;
+    };
+
+    /// \brief To speed up calculation of kernel values we provide a couple of lookup
+    /// tables.
+    ///
+    /// lookup_ is a table of lookups. lookup_[0] is the kernel
+    /// value at a distance of 0, and lookup_.back() is the kernel value at the
+    /// largest possible distance (diagonal across the bounding box).
```
Comment (matt-gretton-dann) [inline src/Kernels.h:67]: `(diagonal across the bounding box)` is incorrect with spherical co-ordinates.  See https://github.com/mrc-ide/covid-sim/blob/master/src/SetupModel.cpp#L201

Comment (zebmason) [reply src/Kernels.h:67]: Bit obvious I suppose - the bounding box could straddle the equator and we do live on an oblate spheroid.

**Comment by zebmason** [comment]
@matt-gretton-dann I did think this would just be a first commit to move the code around and thus be functionally identical. I had intended to re-do the numerics of the table calculation as they waste a lot of flops and I didn't see the compiler ever being intelligent enough to optimise them but I suppose they're not important if the table only needs calculating the once.

**Comment by weshinsley** [comment]
I don't know why the CI for macOS is doing this - no-one has cancelled it...

**Comment by zebmason** [comment]
@weshinsley It used to be that one of the Linux containers used to time out. Rather a pain but I usually just amend the last commit to create a new hash and then force push. Have a good weekend.

**Comment by zebmason** [comment]
Unit tests look a tough one. The actual kernel functions can be tested with a few different values but they are trivial apart from discerning appropriate values. I'd kick it back to Imperial and ask them to think on the tests.

**Comment by zebmason** [comment]
@weshinsley, @zdroid Interestingly this PR had already implemented all the changes in #419 some time ago

**Comment by weshinsley** [comment]
Yeah, I know... PR #419 was a fairly easy one for me to review...! This one may be better for Matt to finish off...

**Pull request merged**